network/iptables.go

```diff
@@ -16,145 +16,273 @@
-	}()
+	for _, rule := range c.legacyRules(ipt) {
+		glog.Info("Deleting legacy iptables rule: table: ", rule.table, " chain: ", rule.chain, " spec: ",
+			strings.Join(rule.getRule(), " "))
 
-	for {
-		// Ensure that all the iptables rules exist every 5 seconds
-		if err := ensureIPTables(ipt, rules); err != nil {
-			log.Errorf("Failed to ensure iptables rules: %v", err)
-		}
-
-		time.Sleep(time.Duration(resyncPeriod) * time.Second)
+		// legacy rules are likely not on the system other than rare circumstances - ignore errors
+		_ = ipt.Delete(rule.table, rule.chain, rule.getRule()...)
 	}
 }
 
-func ensureIPTables(ipt IPTables, rules []IPTablesRule) error {
-	exists, err := ipTablesRulesExist(ipt, rules)
-	if err != nil {
-		return fmt.Errorf("Error checking rule existence: %v", err)
-	}
-	if exists {
-		// if all the rules already exist, no need to do anything
-		return nil
+func (c Config) createRules(ipt IPTables) error {
+	for _, chain := range chains {
+		err := ipt.ClearChain(chain.table, chain.name)
+		if err != nil {
+			glog.Info("Clear chain ", chain, " failed: ", err)
+		}
 	}
-	// Otherwise, teardown all the rules and set them up again
-	// We do this because the order of the rules is important
-	log.Info("Some iptables rules are missing; deleting and recreating rules")
-	teardownIPTables(ipt, rules)
-	if err = setupIPTables(ipt, rules); err != nil {
-		return fmt.Errorf("Error setting up rules: %v", err)
+
+	for _, rule := range c.generateRules(ipt, c.Masquerade) {
+		glog.Info("Adding iptables rule: table: ", rule.table, " chain: ", rule.chain, " spec: ",
+			strings.Join(rule.getRule(), " "))
+
+		err := ipt.AppendUnique(rule.table, rule.chain, rule.getRule()...)
+		if err != nil {
+			return err
+		}
 	}
 	return nil
 }
 
-func setupIPTables(ipt IPTables, rules []IPTablesRule) error {
-	for _, rule := range rules {
-		log.Info("Adding iptables rule: ", strings.Join(rule.rulespec, " "))
-		err := ipt.AppendUnique(rule.table, rule.chain, rule.rulespec...)
+func (c Config) rulesOk(ipt IPTables) error {
+	for _, rule := range c.generateRules(ipt, c.Masquerade) {
+		exists, err := ipt.Exists(rule.table, rule.chain, rule.getRule()...)
 		if err != nil {
-			return fmt.Errorf("failed to insert IPTables rule: %v", err)
+			return trace.Wrap(err)
+		}
+		if !exists {
+			return trace.NotFound("missing rule")
 		}
 	}
-
 	return nil
 }
 
-func teardownIPTables(ipt IPTables, rules []IPTablesRule) {
-	for _, rule := range rules {
-		log.Info("Deleting iptables rule: ", strings.Join(rule.rulespec, " "))
-		// We ignore errors here because if there's an error it's almost certainly because the rule
-		// doesn't exist, which is fine (we don't need to delete rules that don't exist)
-		ipt.Delete(rule.table, rule.chain, rule.rulespec...)
+func (c Config) ensureRules(ipt IPTables) {
+	err := c.rulesOk(ipt)
+	if err != nil && !trace.IsNotFound(err) {
+		glog.Error("Error checking iptables rules: ", trace.DebugReport(err))
```

Comment:
Is the stackl trace going to be readable in this case?

Reply:
May need some sedding, but I don't think we want to rip out glog and replace with logrus across the entire project in this PR.